### PR TITLE
Add no local GEMM partial vectorization and selectable vectorization types

### PR DESCRIPTION
**Pull request by CodeplayBen**
Adds the first vectorized version of the no_local GEMM kernel as a seperate file, also adds an enum to GEMM configurations to allow selecting the type of vectorization to use. This allows us more options for picking the best kernel configuration for different backends.

This partial vectorization of the no_local kernel is only vectorizable when A is not transposed and B is transposed.

### CODE REVIEW

File: src/operations/blas3/gemm_no_local_partial_vec.hpp [hunk 1]
```diff
@@ -0,0 +1,564 @@
...
+    orig_A += dim_m_a_start * (trans_a ? lda : 1);
+    orig_B += dim_n_b_start * (trans_b ? 1 : ldb);
+    orig_C += dim_m_a_start + (dim_n_b_start * ldc);
+
+    /*
+     * The following lambdas: boundary_check_m, boundary_check_n, and
+     * boundary_check_c  are used to check the A, B , and C boundaries
+     * respectively.
+     */
+    const auto boundary_check_m = [&](index_t dim_m_a_start) {
```
Comment (mehdi-goli) [inline src/operations/blas3/gemm_no_local_partial_vec.hpp]: Add SYCL_BLAS_ALWAYS_INLINE -> ` [&](index_t dim_m_a_start) SYCL_BLAS_ALWAYS_INLINE { `

File: src/operations/blas3/gemm_no_local_partial_vec.hpp [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+      return dim_m_a_start < m;
+    };
+    const auto boundary_check_n = [&](index_t dim_n_b_start) {
```
Comment (mehdi-goli) [inline src/operations/blas3/gemm_no_local_partial_vec.hpp]: same as above

File: src/operations/blas3/gemm_no_local_partial_vec.hpp [hunk 3]
```diff
@@ -0,0 +1,564 @@
...
+      return dim_n_b_start < n;
+    };
+    const auto boundary_check_c = [&](index_t dim_m_c_start,
```
Comment (mehdi-goli) [inline src/operations/blas3/gemm_no_local_partial_vec.hpp]: same as above

Comment (CodeplayBen) [reply src/operations/blas3/gemm_no_local_partial_vec.hpp]: Fixed!

File: src/operations/blas3/gemm_no_local_partial_vec.hpp (hunk 2 above)
Comment (CodeplayBen) [reply src/operations/blas3/gemm_no_local_partial_vec.hpp]: Fixed!

File: src/operations/blas3/gemm_no_local_partial_vec.hpp (hunk 1 above)
Comment (CodeplayBen) [reply src/operations/blas3/gemm_no_local_partial_vec.hpp]: Fixed!

File: src/operations/blas3/gemm_load_store.hpp [hunk 4]
```diff
@@ -69,51 +71,49 @@ struct Packetize {
    * @tparam internal True if the current block is internal and no bounds
    * checking is required.
    * @tparam ld The leading dimension of the destination memory. */
-  template <bool trans, bool internal, int ld, typename packet_t = PacketType,
-            typename SrcPointerType, typename DestPointerType, typename index_t,
-            typename EdgePredicate>
+  template <bool trans, bool internal, index_t ld,
+            typename packet_t = PacketType, typename SrcPointerType,
```
Comment (jwlawson) [inline src/operations/blas3/gemm_load_store.hpp]: Was `packet_t` ever not `PacketType`? I thought you'd removed this parameter (but maybe I'm remembering a change from somewhere else).

File: src/operations/blas3/gemm_local.hpp [hunk 5]
```diff
@@ -234,6 +231,7 @@ class Gemm<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize, TileType,
   template <typename local_memory_t>
   SYCL_BLAS_INLINE void eval(local_memory_t scratch_acc,
                              const cl::sycl::nd_item<1> &id) noexcept {
+    printf("local memory full vec\n");
```
Comment (jwlawson) [inline src/operations/blas3/gemm_local.hpp]: This should probably be removed

File: src/operations/blas3/gemm_load_store.hpp (hunk 4 above)
Comment (CodeplayBen) [reply src/operations/blas3/gemm_load_store.hpp]: Ah yes, that change is in the subsequent PR I'll be putting up!

File: src/operations/blas3/gemm_local.hpp (hunk 5 above)
Comment (CodeplayBen) [reply src/operations/blas3/gemm_local.hpp]: Fixed, thanks!

**Pull request merged**